plinky/src/passes/gc_sections.rs

```diff
@@ -33,16 +33,21 @@ pub(crate) fn run(object: &mut Object) -> Vec<RemovedSection> {
         visitor.add(symbol.id());
     }
 
-    // Mark all sections that will not be allocated in memory to be saved, as checking the
-    // relocations from the entry point is not accurate for that.
     for section in object.sections.iter() {
+        // Mark all sections that will not be allocated in memory to be saved, as checking the
+        // relocations from the entry point is not accurate for that.
         match &section.content {
             SectionContent::Data(data) if data.perms.read => {}
             SectionContent::Uninitialized(uninit) if uninit.perms.read => {}
             _ => {
                 visitor.to_save.insert(section.id);
             }
         }
+
+        // Mark all retained (`#[used(linker)`, `__attribute__((retain))`) section as saved.
+        if section.retain {
+            visitor.to_save.insert(section.id);
```

Comment:
We want to add the section to the visitor queue, instead of adding it to the list of sections to save. Otherwise, if this is a data section with relocations, the symbols the relocations point to might be garbage collected.

I see this is also a problem above on line 43 :sweat_smile: